### PR TITLE
feat(argo-cd): Support Azure DevOps webhook Secret

### PR DESCRIPTION
**Pull request by yu-croco**
Resolves https://github.com/argoproj/argo-helm/issues/2432 .

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

### CODE REVIEW

File: charts/argo-cd/values.yaml [hunk 1]
```diff
@@ -516,6 +516,12 @@ configs:
     bitbucketUUID: ""
     # -- Shared secret for authenticating Gogs webhook events
     gogsSecret: ""
+    ## Azure DevOps
+    azureDevops:
+      # -- Shared secret username for authenticating Azure DevOps webhook events
+      username: ""
+      # -- Shared secret password for authenticating Azure DevOps webhook events
+      password: ""
```
Comment (yu-croco) [inline charts/argo-cd/values.yaml:524]: 📝 
Ref: https://argo-cd.readthedocs.io/en/stable/operator-manual/webhook/#2-configure-argo-cd-with-the-webhook-secret-optional

image

File: charts/argo-cd/templates/argocd-configs/argocd-secret.yaml [hunk 2]
```diff
@@ -34,6 +34,10 @@ data:
   {{- with .Values.configs.secret.gogsSecret }}
   webhook.gogs.secret: {{ . | b64enc }}
   {{- end }}
+  {{- with .Values.configs.secret.azureDevops }}
+  webhook.azuredevops.username: {{ .username }}
+  webhook.azuredevops.password: {{ .password | b64enc }}
+  {{- end }}
```
Comment (yu-croco) [inline charts/argo-cd/templates/argocd-configs/argocd-secret.yaml:40]: 📝 
Ref: https://argo-cd.readthedocs.io/en/stable/operator-manual/webhook/#2-configure-argo-cd-with-the-webhook-secret-optional

image

**Review by jmeridth — APPROVED** [review]

**Review by mbevc1 — APPROVED** [review]

**Comment by TechDawg270** [comment]
not sure why but this introduced a breaking change with error message:
`Error: Failed to render chart: exit status 1: Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: [unknown object type "nil" in Secret.data.webhook.azuredevops.password, unknown object type "nil" in Secret.data.webhook.azuredevops.username]`
chart version `5.53.3` works fine however. That was generated running helm with the diff plugin

**Comment by yu-croco** [comment]
Hi @TechDawg270 , thank you for the info.
Can you please share us your values.yaml and version of Helm ?

CI in this PR has been passed, so the default value.yaml is supposed to be harmless. 🤔

**Comment by cherrot** [comment]
Maybe we should check whether `.Values.configs.secret.azureDevops` is `nil` before using it?

I can reproduce @TechDawg270 's issue via the following values and command:

```yaml
  secret:
    gitlabSecret: ""
    extra:
      oidc.clientSecret: ""
```

```shell
helm upgrade --install argocd -n argocd argo/argo-cd \
    -f my-values.yaml \
    --set configs.secret.extra.'oidc\.clientSecret'=${MY_OIDC_SECRET}  --dry-run
```

**Comment by yu-croco** [comment]
Can you share your version of Helm ?

**Comment by cherrot** [comment]
> Can you share your version of Helm ?

```shell
helm version
version.BuildInfo{Version:"v3.13.3", GitCommit:"c8b948945e52abba22ff885446a1486cb5fd3474", GitTreeState:"clean", GoVersion:"go1.21.5"}
```

**Comment by yu-croco** [comment]
Ah sorry, iOS app didn't show full message...

The following version worked well, so it seems this is due to older Helm version.
I will add condition to  check whether .Values.configs.secret.azureDevops is nil for older Helm version.

```
version.BuildInfo{Version:"v3.14.0", GitCommit:"3fc9f4b2638e76f26739cd77c7017139be81d0ea", GitTreeState:"clean", GoVersion:"go1.21.6"}
```

**Comment by cherrot** [comment]
Wait a second! Helm `v3.14.0` doesn't solve this issue on my environment.  I'm trying to narrow down the minimal reproduce set.

Thanks for quick responding :heart:

**Comment by yu-croco** [comment]
Thank you for your cooperation!

> Helm v3.14.0 doesn't solve this issue on my environment

😢

**Comment by cherrot** [comment]
To reproduce on my Arch Linux:

My values file: `cat deploy/argocd/values-prod.yaml`

```yaml
configs:
  secret:
    gitlabSecret: 'any'
```

Command: 

```shell
/tmp/helm install argocd -n test argo/argo-cd -f deploy/argocd/values-prod.yaml --version=5.53.4 --dry-run

```

Got: 

```
Error: INSTALLATION FAILED: unable to build kubernetes objects from release manifest: error validating "": error validating data: [unknown object type "nil" in Secret.data.webhook.azuredevops.password, unknown object type "nil" in Secret.data.webhook.azuredevops.username]
```

I think `Helm` acts like simply overwriting the whole object (`configs.secret` in this case), which causes `configs.secret.azureDevops` being `nil`.

**Comment by yu-croco** [comment]
Thank you for the info.
Hmm, it works in my local well and the CI in this PR works... I have no idea...

Do you think the PR below helps you ? 🤔 (I will fix the comment)
https://github.com/argoproj/argo-helm/pull/2443/files

**Comment by yu-croco** [comment]
ahhh now i recognize the issue.

**Comment by yu-croco** [comment]
Anyway, I will fix in https://github.com/argoproj/argo-helm/pull/2443 . 🙏

**Pull request merged**